File: HW1/hw1-handout/hw1-handout/matrix/matrix.go
package matrix

// If needed, you may define helper functions here.

// AreAdjacent returns true iff a and b are adjacent in lst.
func AreAdjacent(lst []int, a, b int) bool {
	if lst == nil || len(lst) == 0 {
		return false
	}
	for i := 0; i < len(lst); i++ {
		if lst[i] == a {
			if i-1 >= 0 && lst[i-1] == b {
				return true
			} else if i+1 < len(lst) && lst[i+1] == b {
				return true
			}
		}
	}
	return false
}

// Transpose returns the transpose of the 2D matrix mat.
func Transpose(mat [][]int) [][]int {
	if mat == nil || len(mat) == 0 || len(mat[0]) == 0 {
		return mat
	}
	row := len(mat)
	col := len(mat[0])
	res := make([][]int, col)
	for i := 0; i < col; i++ {
		res[i] = make([]int, row)
	}
	for i := 0; i < col; i++ {
		for j := 0; j < row; j++ {
			res[i][j] = mat[j][i]
		}
	}
	return res
}

// AreNeighbors returns true iff a and b are Manhattan neighbors in the 2D
// matrix mat.
func AreNeighbors(mat [][]int, a, b int) bool {
	for i := 0; i < len(mat); i++ {
		for j := 0; j < len(mat[i]); j++ {
			if mat[i][j] == a {
				if i-1 >= 0 && mat[i-1][j] == b {
					return true
				} else if i+1 < len(mat) && mat[i+1][j] == b {
					return true
				} else if j-1 >= 0 && mat[i][j-1] == b {
					return true
				} else if j+1 < len(mat[i]) && mat[i][j+1] == b {
					return true
				}
			}
		}
	}
	return false
}

File: HW4/hw4-handout/hw4-handout/matrix/matrix.pl
% A list is a 1-D array of numbers.
% A matrix is a 2-D array of numbers, stored in row-major order.

% You may define helper functions here.

% are_adjacent helper functions:
match_first_element([H|_], H).

are_adjacent_recursion([A|T], A, B) :-
    match_first_element(T, B)
.

are_adjacent_recursion([B|T], A, B) :-
    match_first_element(T, A)
.

are_adjacent_recursion([_|T], A, B) :-
    are_adjacent_recursion(T, A, B)
.

% matrix_transpose helper functions:
matrix_transpose_recursion([], []).
matrix_transpose_recursion([[]|_], []).
matrix_transpose_recursion(Matrix, Answer) :-
    construct_row(Matrix, MatrixNoFirstColumn, FirstAnswerRow),
    matrix_transpose_recursion(MatrixNoFirstColumn, SubAnswer),
    append_to_matrix(FirstAnswerRow, SubAnswer, Answer)
.

construct_row([], [], []).
construct_row([H|T], MatrixNoFirstColumn, Row) :-
    first_element(H, First),
    rest_elements(H, Rest),
    construct_row(T, SubMatrixNoFirstColumn, SubRow),
    append_to_matrix(Rest, SubMatrixNoFirstColumn, MatrixNoFirstColumn),
    append(First, SubRow, Row)
.

first_element([H|_], [H|[]]).

rest_elements([_|T], T).

append_to_matrix(Row, M, [Row|M]).

% are_neighbors helper functions:
are_not_neighbors(Matrix, A, B) :-
    matrix_transpose(Matrix, Answer),
    are_not_neighbors_recursion(Matrix, A, B),
    are_not_neighbors_recursion(Answer, A, B)
.

are_not_neighbors_recursion([], _, _) :- true.
are_not_neighbors_recursion([H|T], A, B) :-
    not(are_adjacent(H, A, B)),
    are_not_neighbors_recursion(T, A, B)
.

% are_adjacent(List, A, B) returns true iff A and B are neighbors in List.
are_adjacent(List, A, B) :-
    are_adjacent_recursion(List, A, B)
.

% matrix_transpose(Matrix, Answer) returns true iff Answer is the transpose of
% the 2D matrix Matrix.
matrix_transpose(Matrix, Answer) :-
    matrix_transpose_recursion(Matrix, Answer)
.

% are_neighbors(Matrix, A, B) returns true iff A and B are neighbors in the 2D
% matrix Matrix.
are_neighbors(Matrix, A, B) :-
    not(are_not_neighbors(Matrix, A, B))
.